extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -76,21 +77,52 @@
                     });
         }
 
-        final String code = context.request().getParam("code");
-        if (code == null) {
-            return Uni.createFrom().optional(Optional.empty());
+        final Cookie stateCookie = context.request().getCookie(getStateCookieName(oidcTenantConfig));
```

Comment:
This looks fine, but my opinion is still that this should be conditional to the current request being a callback URI and not just _any_ URI. That's what I do in my branch. What's the reason you don't check that?

Reply:
@FroMage Let me thinks about it

Reply:
@FroMage Ok, I don't see it making any difference in which order to start checking it as both actions: recalculating the redirect URI and checking if the state cookie exists and valid, have to be done to complete the code exchange (the redirect uri has to be sent with the code grant alongside the code).
 
If it an initial non-authenticated request then it is no more expensive to start with checking if the state cookie with the given name exists (which means it is a redirect which must've come from the provider) than starting with recalculating the absolute redirect URI and comparing it with the current absolute one, even a bit cheaper I'd say.

I think it is safer to start from checking the state cookie, to avoid breaking some user authentications where the recalculated and redirect uris differ in some trailing `/` - because if we start comparing them then we'd need to do what the OIDC spec requires, do the exact match, while some providers might be lenient enough to handle these minor differences and they have to compare it anyway when the code to token exchange is requested.

All in all, IMHO, it is better to leave it as is. It is not going affect the treatment of the non-authenticated requests with `code`, `error`, etc. if this code fails where we have a state cookie, `code` but it is not a redirect from OIDC provider then it will mean there is a bug in removing the state cookies - but I don't think it is going to happen, we test the cookies are being removed

Reply:
@FroMage  I nearly started coding again :-) around replacing the referrer check with the redirect uri check, but again, I don't think it will work reliably, apart from the minor `/` differences, the issue with forcing `https`, we have to compare the absolute URI, it is just too messy, i.e Routing Context provides the `http:` based one, we calculate it as `https:` based one, etc.
I'll just leave things as they are

Reply:
Well, it's not an issue of speed but correctness: if we have a state cookie but we land for some reason on a page which is not a callback URI, the OIDC code will be invoked and read the `code` param (which may not be related to OIDC) and possibly read the form body when it shouldn't. I just don't see why we even look at anything if we're not in a callback URI.

Reply:
@FroMage I know, my comments can be long, but please read through them, I've gone through all of it; the speed was only mentioned to support the fact the state cookie is checked first, but both the cookie and the redirect URI have to be checked.

>  if we have a state cookie but we land for some reason on a page which is not a callback URI

`quarkus-oidc` does not have uncertainties like this one. It won't happen. As I said above, if it does happen - then it has a bug but we do test the state cookies are removed - they are removed immediately after a redirect. So a case where an unauthenticated user calls with a custom `code` but the state cookie is there is theoretical and does not apply to quarkus-oidc.

>  I just don't see why we even look at anything if we're not in a callback URI.

We do calculate it to complete the code exchange if the state cookie exists. If the state cookie exists then it must mean only one thing - it is a redirect from OIDC
 
if you'd like you can test this PR to verify.

Reply:
OK whatever.

Reply:
@FroMage Have a look at `CodeAuthenticationMechanism` in this PR, the `absolute` redirect URI is recalculated pretty much immediately after the state cookie is checked - it is submitted to the code to token exchange request - it will fail if it does not match the absolute  one registered in the provider. See also my concerns on doing the comparing the redirect URIs in quarkus-oidc - it can be messy and not very reliable - we let the provider compare it which is required in any case.

I think the current approach is sound, it has really worked well until now so I don't think we should start changing it. And this PR fixes the `code` bug in the unauthenticated request - and the `form_post` support will flow from it

Reply:
@FroMage So if you also check the redirect URI in your branch then you basically do it twice in a row (and if you check the relative paths then it is not reliable)